### PR TITLE
Perf tracker

### PR DESCRIPTION
**Pull request by nzmora**
Introduce a performance tracking class

This will help define and use different performance sorting schemes.
E.g. this will address the issue raised in issue #411

### CODE REVIEW

File: distiller/apputils/performance_tracker.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# limitations under the License.
+#
+
+"""Performance trackers used to track the best performing epochs when training.
+"""
+import operator
+import distiller
+
+
+__all__ = ["SparsityAccuracyTracker"]
```
Comment (barrh) [inline distiller/apputils/performance_tracker.py]: Can you expose both?
I would prefer customizing in classifier.py than here.

Comment (nzmora) [reply distiller/apputils/performance_tracker.py]: done; But the reason I didn't expose it in the first place is that I think that we can all benefit from sub-classes (for handling ranking based on number of quantized weights and activations) - so think of that when you decide where to place your sub-classes.

File: distiller/apputils/performance_tracker.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+class AccuracyPerformanceTracker(object):
+    """Base class for performance trackers using Top1 and Top5 accuracy metrics"""
+    def __init__(self, num_best_scores):
+        self.perf_scores_history = []
+        self.max_len = num_best_scores
+
+    def reset(self):
+        self.perf_scores_history = []
+
+    def step(self, model, top1, top5, epoch):
```
Comment (guyjacob) [inline distiller/apputils/performance_tracker.py]: Maybe use a list of score-names and score-values pairs instead of hard coded top-1/5? I mean, one could always "abuse" this in non-classification tasks and pass different metrics, but that's not intuitive.

File: distiller/apputils/performance_tracker.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        """Update the list of top training scores achieved so far"""
+        raise NotImplementedError
```
Comment (guyjacob) [inline distiller/apputils/performance_tracker.py]: Why not have a default implementation using just accuracy values?

File: distiller/apputils/image_classifier.py [hunk 4]
```diff
@@ -76,6 +75,7 @@ def __init__(self, args, script_dir):
         self.train_loader, self.val_loader, self.test_loader = (None, None, None)
         self.activations_collectors = create_activation_stats_collectors(
             self.model, *self.args.activation_stats)
+        self.performance_tracker = apputils.SparsityAccuracyTracker(self.args.num_best_scores)
```
Comment (guyjacob) [inline distiller/apputils/image_classifier.py]: I don't think this should be the default - you'd get the same problem as #411. Assuming that was the motivation for this, maybe defer the selection of the tracker until a compression scheduler was created, and only if pruning/regularization policies exist then use this tracker. This also assumes a default tracker that only looks at accuracy (see comment below).
(I realize that the way you implemented it current functionality - so this suggestion can be deferred to a later PR if you want)

File: distiller/apputils/performance_tracker.py (hunk 2 above)
Comment (barrh) [reply distiller/apputils/performance_tracker.py]: @guyjacob I agree that enabling flexible API can benefit the user, but in this case, I feel that it has the potential to over-complicate the API for limited benefit.

e.g. If I were to come up with alternative params for ranking it would be: `(top1, loss, epoch)`
- top1 - as before
- loss - Distinguished from top5 due to lower-is-better attribute, and has to scope of the entire universe of classes.
- epoch - for stable ordering

To allow to above params, the "pairs" API should first be upgraded to "trios" API: (name,value,asc/desc).
For conclusion, pairs don't benefit me, but trios do, then again - complicated.

File: distiller/apputils/performance_tracker.py (hunk 3 above)
Comment (nzmora) [reply distiller/apputils/performance_tracker.py]: @guyjacob - please review the new API.  It is now more generic.

File: distiller/apputils/performance_tracker.py (hunk 2 above)
Comment (nzmora) [reply distiller/apputils/performance_tracker.py]: @guyjacob, @barrh  - please review the new API.  It is now more generic.
@barrh - to control the "direction" of the ranking (what you refer to as asc/desc above) just pass the score multiplied by -1.
For example, in `SparsityAccuracyTracker` I want to prefer lower parameter NNZ count, so I sort by `-params_nnz_cnt`.

Comment (barrh) [reply distiller/apputils/performance_tracker.py]: @nzmora Wouldn't it reflect in the logging as negative value?

Comment (nzmora) [reply distiller/apputils/performance_tracker.py]: @barrh the logging code is orthogonal and handles this:
`self.performance_tracker.step(self.model, epoch, top1=top1, top5=top5)
        _log_best_scores(self.performance_tracker, msglogger)
`

`_log_best_scores` is a private function of `image_classifier.py` and is currently specific to `apputils.SparsityAccuracyTracker` (I just pushed a commit to make this explicit in code).  If/when someone uses a different PerformanceTracker, they can generalize this logging function.
